modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java

```diff
@@ -678,8 +704,17 @@ private Definition(List<Whitelist> whitelists) {
     }
 
     private void addStruct(ClassLoader whitelistClassLoader, Whitelist.Struct whitelistStruct) {
-        if (!whitelistStruct.painlessTypeName.matches("^[_a-zA-Z][._a-zA-Z0-9]*")) {
-            throw new IllegalArgumentException("invalid struct type name [" + whitelistStruct.painlessTypeName + "]");
+        String painlessTypeName = whitelistStruct.javaClassName.replace('$', '.');
+        String importedPainlessTypeName = painlessTypeName;
+
+        if (painlessTypeName.matches("^[_a-zA-Z][._a-zA-Z0-9]*") == false) {
```

Comment:
Maybe move this out to a private static constant so the regex is not recompiled for every struct in the whitelist?

Reply:
Done.